pkg/common/metrics/stats_reporter.go

```diff
@@ -0,0 +1,207 @@
+// This wrapper function facilitates minimally-invasive unit testing of the
+// Report functionality without requiring live servers to be started.
+var RecordWrapper = metrics.Record
+
+// Some type aliases for the otherwise unwieldy metric collection map-of-maps-to-interfaces
+type ReportingItem = map[string]interface{}
+type ReportingList = map[string]ReportingItem
+
+//
+// Report The Sarama Metrics (go-metrics) Via Knative / OpenCensus Metrics
+//
+func (r *Reporter) Report(list ReportingList) {
+
+	// Validate The Metrics
+	if len(list) > 0 {
+
+		// Loop Over The Observed Metrics
+		for metricKey, metricValue := range list {
+
+			// Record Each Individual Metric Item
+			for saramaKey, saramaValue := range metricValue {
+				r.recordMeasurement(metricKey, saramaKey, saramaValue)
+			}
+		}
+	}
+}
+
+// Creates and registers a new view in the OpenCensus context, adding it to the Reporter's known views
+func (r *Reporter) createView(ctx context.Context, measure stats.Measure, name string, description string) context.Context {
+	key, err := tag.NewKey(name)
+	if err != nil {
+		r.logger.Error("Failed To Create New OpenCensus Key For Sarama Metric", zap.String("Key", name), zap.Error(err))
+		return ctx
+	}
+	ctx, err = tag.New(ctx, tag.Insert(key, name))
+	if err != nil {
+		r.logger.Error("Failed To Create New OpenCensus Tag For Sarama Metric", zap.String("Key", name), zap.Error(err))
+		return ctx
+	}
+
+	newView := &view.View{
+		Name:        name,
+		Description: description,
+		Measure:     measure,
+		Aggregation: view.LastValue(),
+	}
+	err = view.Register(newView)
+	if err != nil {
+		r.logger.Error("failed to register opencensus views", zap.Error(err))
+		return ctx
+	}
+	r.views[name] = newView
+	return ctx
+}
+
+// Record a measurement to the metrics backend, creating a new OpenCensus view if this is a new measurement
+func (r *Reporter) recordMeasurement(metricKey string, saramaKey string, value interface{}) {
+
+	name := fmt.Sprintf("%s.%s", metricKey, saramaKey)
+	description := getDescription(metricKey, saramaKey)
+
+	// Type-switches don't support "fallthrough" so each individual possible type must have its own
+	// somewhat-redundant code block
+	switch value := value.(type) {
+	case int:     r.recordInt(int64(value), name, description)
+	case int8:    r.recordInt(int64(value), name, description)
+	case int16:   r.recordInt(int64(value), name, description)
+	case int32:   r.recordInt(int64(value), name, description)
+	case int64:   r.recordInt(value, name, description)
+	case uint:    r.recordInt(int64(value), name, description)
+	case uint8:   r.recordInt(int64(value), name, description)
+	case uint16:  r.recordInt(int64(value), name, description)
+	case uint32:  r.recordInt(int64(value), name, description)
+	case uint64:  r.recordInt(int64(value), name, description)
+	case float64: r.recordFloat(value, name, description)
+	case float32: r.recordFloat(float64(value), name, description)
+	default:
```

Comment:
I've removed most of the unused types to make this simpler, but you can't exactly do what you've suggested there, because if you combine values in a type switch, you get "interface{}" as the value again (instead of "int32" or "int64" or etc.), which is exactly what I was trying to get away from by doing the switch in the first place (and then the int64(value) is invalid code).